kong/mlcache.lua

```diff
@@ -0,0 +1,466 @@
+
+    [4] = function(t)      -- table
+        local json, err = cjson.encode(t)
+        if not json then
+            return nil, "could not encode table value: " .. err
+        end
+
+        return json
+    end,
+}
+
+
+local unmarshallers = {
+    shm_value = function(marshalled)
+        local ttl_last = find(marshalled, ":", 21, true) - 1
+
+        local value_type = sub(marshalled, 1, 1)
+        local at         = sub(marshalled, 3, 19)
+        local ttl        = sub(marshalled, 21, ttl_last)
+        local str_value  = sub(marshalled, ttl_last + 2)
+
+        return str_value, tonumber(value_type), tonumber(at), tonumber(ttl)
+    end,
+
+    [1] = function(str) -- number
+        return tonumber(str)
+    end,
+
+    [2] = function(str) -- boolean
+        return str == "true"
+    end,
+
+    [3] = function(str) -- string
+        return str
+    end,
+
+    [4] = function(str) -- table
+        local t, err = cjson.decode(str)
+        if not t then
+            return nil, "could not decode table value: " .. err
+        end
+
+        return t
+    end,
+}
+
+
+local _M = {}
+local mt = { __index = _M }
+
+
+function _M.new(shm, ipc, opts)
+    if type(shm) ~= "string" then
+        return error("shm must be a string")
+    end
+
+    if not ipc then
+        return error("ipc must be given")
+    end
+
+    if opts then
+        if type(opts) ~= "table" then
+            return error("opts must be a table")
+        end
+
+        if opts.lru_size and type(opts.lru_size) ~= "number" then
+            return error("opts.lru_size must be a number")
+        end
+
+        if opts.ttl then
+            if type(opts.ttl) ~= "number" then
+                return error("opts.ttl must be a number")
+            end
+
+            if opts.ttl < 0 then
+                return error("opts.ttl must be >= 0")
+            end
+        end
+
+        if opts.neg_ttl then
+            if type(opts.neg_ttl) ~= "number" then
+                return error("opts.neg_ttl must be a number")
+            end
+
+            if opts.neg_ttl < 0 then
+                return error("opts.neg_ttl must be >= 0")
+            end
+        end
+
+    else
+        opts = {}
+    end
+
+    local dict = shared[shm]
+    if not dict then
+        return nil, "no such lua_shared_dict: " .. shm
+    end
+
+    local lru_size = opts.lru_size or 100
+
+    local self   = {
+        lru_size = lru_size,
+        lru      = lrucache.new(lru_size),
+        dict     = dict,
+        shm      = shm,
+        ttl      = opts.ttl     or 30,
+        neg_ttl  = opts.neg_ttl or 5
+    }
+
+    ipc.register(function(data)
+        local key = data
+
+        self.lru:delete(key)
+    end, "mlcache", "invalidations")
+
+    ipc.register(function()
+        self.lru = lrucache.new(self.lru_size)
+    end, "mlcache", "purge")
+
+    self.ipc = ipc
+
+    return setmetatable(self, mt)
+end
+
+
+function _M:purge()
+    self.dict:flush_all()
+    self.dict:flush_expired()
```

Comment:
A boolean value could be implemented here, that'd be neat, and yes, using the LRU mechanism is definitely fine I think